Fuzzy finder: refactor and implement Figma design

olafurpg opened the pull request:
This is the first step towards implementing the design from https://www.figma.com/file/E0vD9UduSp2liVyDb7dplw/Fuzzy-finder-UI-improvements?node-id=1714%3A3559

While working on the code, I was increasingly unhappy with the convoluted state management inside `useFuzzyState`, which has caused app freezes (infinite loops between cyclic hooks).

The new state management inside `useFuzzyState` is significantly simpler to reason about and it should be much easier to add new functionality to this new codestructure.

The "Actions" tab is no longer enabled with `fuzzyFinderAll` since I suspect it will take a long time until we can make it GA (need at least a few dozen actions to make it GA-ready).

## Test plan

See the CI go green. To test the changes manually

- `sg start`
- Use keyboard shortcuts to open fuzzy finder
- Type in some query and make sure the fuzzy results are as expected
- Reload the page and make sure the query is preserved and the same results are rendered
- Use tab/shift-tab to cycle between fuzzy tabs
- Use mouse to click on different fuzzy tabs
- Use keyboard shortcut to toggle "search everywhere" and "Search within repo" when focusing on the all/files/symbols fuzzy tabs.
- Use mouse to select "search everywhere" and "search within repo"


<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-olafurpg-fuzzy-v3.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

olafurpg commented:
This PR doesn't implement the full Figma design. The functional improvements are already significant enough so it would be good to merge this before we do the final polishing touches

olafurpg commented:
The UI Review changes are intentional. I made sure that the old behavior remains functionally unchanged (only one shortcut and one hardcoded "Find files" tab) 
 
<img width="1562" alt="CleanShot 2022-10-20 at 20 14 22@2x" src="https://user-images.githubusercontent.com/1408093/197026539-8ce89b44-9d27-4515-ad64-063bae7983ab.png">

limitedmage commented on client/web/src/components/fuzzyFinder/FuzzyFiles.tsx:
```diff
@@ -95,12 +95,70 @@ export function newFuzzyFSM(filenames: string[], createUrl: createUrlFunction):
+        private readonly onNamesChanged: () => void,
+        private readonly repoRevision: FuzzyRepoRevision
+    ) {}
+    public fuzzyFSM(): FuzzyFSM {
```
We should probably add some extra eslint rules for blank lines between functions and CSS rules. WDYT @sourcegraph/frontend-devs ?

limitedmage commented on client/web/src/components/fuzzyFinder/FuzzyModal.tsx:
```diff
@@ -249,50 +322,38 @@ export const FuzzyModal: React.FunctionComponent<React.PropsWithChildren<FuzzyMo
+                    {tabs.isOnlyFilesEnabled() ? (
+                        <H3>Find files</H3>
+                    ) : (
+                        <Tabs
```
When using `Tabs`, the content should be inside the component withing a `TabPanel` for each tab, so that assistive tools know that the content changes when switching tabs. The `Tab`s themselves should also be inside a `TabList`.  See example here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/wildcard/src/components/Tabs/Tabs.story.tsx?L73%3A1-95%3A2=

olafurpg replied:
I added a reminder to follow up on this here https://github.com/sourcegraph/sourcegraph/discussions/42874#discussioncomment-3936374

One concern I have is whether we can correctly use `<TabPanel` since the search results are expensive to compute so it's undesirable to update all tabs on every keystroke 🤔 

Using `<TabList>` should be easy.

olafurpg replied:
I took a stab at using `TabList` and it broke the look for the close button. This will need more work so I'm tracking it with an issue https://github.com/sourcegraph/sourcegraph/issues/43315

limitedmage commented on client/web/src/components/fuzzyFinder/FuzzyQuery.tsx:
```diff
@@ -5,9 +5,11 @@ import { FuzzyFSM } from './FuzzyFsm'
 import { FuzzyLocalCache, PersistableQueryResult } from './FuzzyLocalCache'
```
It looks like this file doesn't need to be `.tsx` (at least I don't see any JSX being used here). To make things easier to navigate, it would be a good idea to go through and rename any files that don't render UI (looks like `FuzzyLocalCache.tsx` is another one?

olafurpg replied:
Opened https://github.com/sourcegraph/sourcegraph/pull/43314

limitedmage reviewed (commented):
This is starting to look really nice! 

In addition to a high-level post-merge code review (didn't go into super high detail with the logic itself), here are some extra usability comments from running the code:

1. The tabs are too close together and it's hard to tell which shortcut belongs to which tab. The figma has more spacing.
![image](https://user-images.githubusercontent.com/206864/197252652-9b833243-6314-45b8-9d3c-85cec0f30254.png)
2. The icons in results are too close to the text. Figma shows more padding. The star icon also needs a bit more left padding.
![image](https://user-images.githubusercontent.com/206864/197252816-cef74081-2e94-453a-9cc4-81d149df473d.png)
3. The highlighted text is too low contrast, at least in dark mode. We need to use different colors in light and dark more, probably `--body-color`.
![image](https://user-images.githubusercontent.com/206864/197252998-f0388395-6fc3-4524-b744-3b92e9c69750.png)
4. The footer also needs some extra top spacing. Right now the search link is way too close to the list and seems like an orphan member of it.
![image](https://user-images.githubusercontent.com/206864/197253399-fee49e0e-af58-4936-a87b-b5daa77ebfc5.png)

olafurpg commented:
@limitedmage Thank you for the review! Me and @quinnkeast paired on the design today and Quinn did a lot of heavy lifting to close more gap between the Figma and the design. The PR is up https://github.com/sourcegraph/sourcegraph/pull/43313 and addresses your points 1, 2, and 4.

Regarding point 3, we're using `--mark-bg` https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/components/fuzzyFinder/HighlightedLink.module.scss?L2&subtree=true
I propose that we fix the CSS variable to have more contrast in dark theme https://sourcegraph.com/github.com/sourcegraph/sourcegraph@olafurpg/fuzzy-styles/-/blob/client/branded/src/global-styles/colors.scss
